components/spi_flash/flash_ops.c

```diff
@@ -127,121 +128,162 @@ esp_err_t IRAM_ATTR spi_flash_erase_range(uint32_t start_addr, uint32_t size)
+     * - Middle part
+     * - Right padding: 4 bytes from the last 4-byte aligned offset covered.
+     */
+    size_t left_off = dst & ~3U;
+    size_t left_size = MIN(((dst + 3) & ~3U) - dst, size);
+    size_t mid_off = left_size;
+    size_t mid_size = (size - left_size) & ~3U;
+    size_t right_off = left_size + mid_size;
+    size_t right_size = (size - mid_size - left_size);
+    rc = spi_flash_unlock();
+    if (rc != SPI_FLASH_RESULT_OK) goto out;
+    if (left_size > 0) {
+        uint32_t t = 0xffffffff;
+        memcpy(((uint8_t *) &t) + (dst - left_off), srcc, left_size);
         spi_flash_disable_interrupts_caches_and_other_cpu();
-        rc = spi_flash_unlock();
-        if (rc == SPI_FLASH_RESULT_OK) {
-            rc = SPIWrite((uint32_t) (dest_addr - pad_head), (const uint32_t*) temp_write_buf, (int32_t) (1024 + pad_head + pad_end));
-            COUNTER_ADD_BYTES(write, 1024 + pad_head + pad_end);
-        }
-        COUNTER_STOP(write);
-        spi_flash_enable_interrupts_caches_and_other_cpu();  
-        if(rc != ESP_OK) {
-            free(temp_write_buf);
-            temp_write_buf = NULL;
-            return spi_flash_translate_rc(rc);
-        }
-
-        free(temp_write_buf);
-        temp_write_buf = NULL;
-        size -= 1024;
-        dest_addr += 1024;
-        src = (uint8_t*)src + 1024;
+        rc = SPIWrite(left_off, &t, 4);
+        spi_flash_enable_interrupts_caches_and_other_cpu();
+        if (rc != SPI_FLASH_RESULT_OK) goto out;
+        COUNTER_ADD_BYTES(write, 4);
     }
-    if(size > 0) {
-        // max need pad byte num for rand size is 6
-        temp_write_buf = (uint8_t*)malloc(size + 6);
-        if(temp_write_buf == NULL) {
-            return ESP_ERR_NO_MEM;
-        }
-        if(dest_addr%4 != 0) {
-            pad_head = dest_addr%4;
-        }
-        if ((pad_head + size)%4 != 0){
-            pad_end = 4 - (pad_head + size) % 4;  
+    if (mid_size > 0) {
+        /* If src buffer is 4-byte aligned as well and is not in a region that
+         * requires cache access to be enabled, we can write it all at once. */
+#ifdef ESP_PLATFORM
+        bool in_dram = ((uintptr_t) srcc >= 0x3FFAE000 &&
+                        (uintptr_t) srcc < 0x40000000);
+#else
+        bool in_dram = true;
+#endif
+        if (in_dram && (((uintptr_t) srcc) + mid_off) % 4 == 0) {
+            spi_flash_disable_interrupts_caches_and_other_cpu();
+            rc = SPIWrite(dst + mid_off, (const uint32_t *) (srcc + mid_off), mid_size);
+            spi_flash_enable_interrupts_caches_and_other_cpu();
+            if (rc != SPI_FLASH_RESULT_OK) goto out;
+            COUNTER_ADD_BYTES(write, mid_size);
+        } else {
+            /*
+             * Otherwise, unlike for read, we cannot manipulate data in the
+             * user-provided buffer, so we write in 32 byte blocks.
+             */
+            while (mid_size > 0) {
+                uint32_t t[8];
```

Comment:
but here we're not doing any padding, this is just the block size. 8 * 4 = 32, just like noted above.